utils.go

```diff
@@ -1,27 +1,122 @@
 package ipfs
 
 import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	ipfs "github.com/ipfs/go-ipfs-api"
+	cmds "github.com/ipfs/go-ipfs-cmds"
+
+	"github.com/beyondstorage/go-endpoint"
+	"github.com/beyondstorage/go-storage/v4/services"
 	"github.com/beyondstorage/go-storage/v4/types"
 )
 
 // Storage is the example client.
 type Storage struct {
+	ipfs *ipfs.Shell
+
 	defaultPairs DefaultStoragePairs
 	features     StorageFeatures
 
+	workDir string
+
 	types.UnimplementedStorager
+	types.UnimplementedCopier
+	types.UnimplementedMover
 }
 
 // String implements Storager.String
 func (s *Storage) String() string {
-	panic("implement me")
+	return fmt.Sprintf("Storager IPFS {WorkDir: %s}", s.workDir)
 }
 
 // NewStorager will create Storager only.
 func NewStorager(pairs ...types.Pair) (types.Storager, error) {
-	panic("implement me")
+	opt, err := parsePairStorageNew(pairs)
+	if err != nil {
+		return nil, err
+	}
+
+	st := &Storage{
+		workDir: "/",
+	}
+	if opt.HasWorkDir {
+		st.workDir = opt.WorkDir
+	}
+
+	ep, err := endpoint.Parse(opt.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	e := "http://localhost:5001"
+	switch ep.Protocol() {
+	case endpoint.ProtocolHTTP:
+		e, _, _ = ep.HTTP()
+	case endpoint.ProtocolHTTPS:
+		e, _, _ = ep.HTTPS()
+	}
+
+	sh := ipfs.NewShell(e)
+	if !sh.IsUp() {
+		return nil, errors.New("ipfs not online")
+	}
+	st.ipfs = sh
+
+	return st, nil
+}
+
+func formatError(err error) error {
+	if _, ok := err.(services.InternalError); ok {
+		return err
+	}
+
+	e, ok := err.(*ipfs.Error)
+	if !ok {
+		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
+	}
+
+	switch e.Message {
+	case "file does not exist", os.ErrNotExist.Error():
+		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)
+	case "internal error":
+		return fmt.Errorf("%w: %v", services.ErrServiceInternal, err)
+	case "rate limited":
+		return fmt.Errorf("%w: %v", services.ErrRequestThrottled, err)
```

Comment:
I think the message is not such a simple one, but will contain more information? BTW, I didn't find instances for these errors 🤔

Reply:
Yes, I didn't find any documentation or examples either.
I checked the relevant code and the `Message` here is indeed not these texts, I will remove these useless branches.